sw/device/lib/dif/dif_alert_handler.h

```diff
@@ -0,0 +1,649 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_ALERT_HANDLER_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_ALERT_HANDLER_H_
+
+/**
+ * @file
+ * @brief <a href="/hw/ip/alert_handler/doc/">Alert Handler</a> Device Interface
+ * Functions
+ */
+
+#include <stddef.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
```

Comment:
Reading through the programmers guide, Initialisation section:
```
...
3. For each class (A..D):

Determine whether to enable escalation mechanisms (accumulation / interrupt timeout) for this particular class. Set CLASSA_CTRL.EN accordingly.

Determine if this class of alerts allows clearing of escalation once it has begun. Set CLASSA_CTRL.LOCK to true if clearing should be disabled. If true, once escalation protocol begins, it can not be stopped, the assumption being that it ends in a chip reset else it will be rendered useless thenceforth.

Determine the number of alerts required to be accumulated before escalation protocol kicks in. Set CLASSA_ACCUM_THRESH accordingly.

Determine whether the interrupt associated with that class needs a timeout. If yes, set CLASSA_TIMEOUT_CYC to an appropriate value greater than zero (zero corresponds to an infinite timeout and disables the mechanism).

For each escalation phase (0..3):

Determine length of each escalation phase by setting CLASSA_PHASE0_CYC accordingly
For each escalation signal (0..3):

Determine whether to enable the escalation signal, and set the CLASSA_CTRL.E0_EN bit accordingly (default is enabled). Note that setting all of the E*_EN bits to 0 within a class has the same effect of disabling the entire class by setting CLASSA_CTRL.EN to zero.
Determine the phase -> escalation mapping of this class and program it via the CLASSA_CTRL.E0_MAP values if it needs to be changed from the default mapping (0->0, 1->1, 2->2, 3->3).
...
```

I wonder if some of the API:
`dif_alert_handler_set_handler_timeout`
`dif_alert_handler_set_phase_duration`
`dif_alert_handler_escalation_counter_get`
...

should be grouped into a single:
`dif_alert_handler_class_configure(const dif_alert_handler_t *alert_handler,
                                                        const dif_alert_handler_class_config_t *config,
                                                         dif_alert_handler_class_t class);`

Something to consider, I guess.

Reply:
That's an interesting idea. @gkelly @lenary what do you think?

I think the main problem is that each class needs to be configured with a list of signals and a list of alerts, the sizes of which will not be known at compile time. I do not want an API that looks like
```
signal_t signals[] = { ... };
config_t config = {
  // ...
  .signals = signals,
};
```
In general, I think configuration APIs that want to accept a variable-length list are awful in C, since they cannot be written as a single constant (though they could be written as a sequence of constants). I think you suggested a similar thing with `bitfield.h` that we decided was a bit too grody.

(I will point out that this would be a non-issue in Rust, because Rust is smart enough to promote the rvalue `[1, 2, 3]` into a constant.)

If you have a concrete API proposal I'm happy to discuss it.

Reply:
> I think the main problem is that each class needs to be configured with a list of signals and a list of alerts, the sizes of which will not be known at compile time

If alert_manager peripheral is similar to PLIC, then it might have relevant register definitions. In PLIC for example, we have `RV_PLIC_PARAM_NUMSRC` and `RV_PLIC_PARAM_NUMTARGET`. The actual named enum with IRQ source IDs, etc then lives in `hw/top_earlgrey/sw/autogen/top_earlgrey.h` for the DIF users benefit.